### PR TITLE
Switch events to eventemitter3 - the saga

### PR DESCRIPTION
**Pull request by acolytec3**
After [failed](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2303) [attempt](https://github.com/ethereumjs/ethereumjs-monorepo/pull/2376) [after](https://github.com/ethereumjs/ethereumjs-monorepo/pull/3330) [attempt](https://github.com/ethereumjs/ethereumjs-monorepo/pull/3735), I'm determined to get this right.

`eventemitter3` is the right choice for us on a performance basis since it's 50% faster than native Nodejs events.

~~Now, I just need to take on the final boss of async events with `eventemitter3`.~~

Final boss has been defeated with the help of the robots.

### CODE REVIEW

File: packages/client/src/sync/fullsync.ts [hunk 1]
```diff
@@ -70,8 +70,8 @@ export class FullSynchronizer extends Synchronizer {
     const syncEvent: Promise<boolean> = new Promise((resolve) => {
       // This event listener listens for other instances of the fetcher that might be syncing from a different peer
       // and reach the head of the chain before the current fetcher.
-      this.config.events.once(Event.SYNC_SYNCHRONIZED, (height?: number) => {
-        this.resolveSync(height)
+      this.config.events.once(Event.SYNC_SYNCHRONIZED, (chainHeight?: bigint) => {
+        this.resolveSync(Number(chainHeight))
```
Comment (jochem-brouwer) [inline packages/client/src/sync/fullsync.ts:74]: We could also propagate the bigint here?

File: packages/client/src/types.ts [hunk 2]
```diff
@@ -56,33 +54,6 @@ export interface EventParams {
   [Event.PROTOCOL_MESSAGE]: [messageDetails: any, protocolName: string, sendingPeer: Peer]
 }
```
Comment (jochem-brouwer) [inline packages/client/src/types.ts:55]: How does the deletion of this event bus change the typing of the client types?

File: packages/client/test/net/protocol/rlpxsender.spec.ts [hunk 3]
```diff
@@ -52,14 +52,3 @@ describe('should receive message', async () => {
...
 })
-
-describe('should catch errors', async () => {
-  const rlpxProtocol = { events: new EventEmitter() }
-  const sender = new RlpxSender(rlpxProtocol as Devp2pETH)
-  it('throws sendStatus error', () => {
-    assert.throws(() => sender.sendStatus({ id: 5 }), /not a function/, 'sendStatus error')
-  })
-  it('throws sendMessage error', () => {
-    assert.throws(() => sender.sendMessage(1, 5), /not a function/, 'sendMessage error')
```
Comment (jochem-brouwer) [inline packages/client/test/net/protocol/rlpxsender.spec.ts:63]: These events are deleted and do not seem related (?) to events, is this intended?

File: packages/client/test/net/protocol/rlpxsender.spec.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ describe('should receive status', async () => {
 
 describe('should receive message', async () => {
   const rlpxProtocol = { events: new EventEmitter() }
-  const sender = new RlpxSender(rlpxProtocol as Devp2pETH)
+  const sender = new RlpxSender(rlpxProtocol as never as Devp2pETH)
```
Comment (jochem-brouwer) [inline packages/client/test/net/protocol/rlpxsender.spec.ts:46]: Why do we have to cast it as `never` first now?

File: packages/evm/README.md [hunk 5]
```diff
@@ -79,9 +79,10 @@ const main = async () => {
   // Note that numbers added are hex values, so '20' would be '32' as decimal e.g.
   const code = [PUSH1, '03', PUSH1, '05', ADD, STOP]
 
-  evm.events.on('step', function (data) {
+  evm.events.on('step', function (data, resolve) {
     // Note that data.stack is not immutable, i.e. it is a reference to the vm's internal stack object
     console.log(`Opcode: ${data.opcode.name}\tStack: ${data.stack}`)
+    resolve?.()
```
Comment (jochem-brouwer) [inline packages/evm/README.md:85]: Isn't the new function sig that `resolve` is always passed? Such that we can just do `resolve()`?

File: packages/evm/src/evm.ts [hunk 6]
```diff
@@ -221,7 +221,16 @@ export class EVM implements EVMInterface {
...
     this._emit = async (topic: string, data: any): Promise<void> => {
-      return new Promise((resolve) => this.events.emit(topic as keyof EVMEvents, data, resolve))
+      const listeners = this.events.listeners(topic as keyof EVMEvent)
+      for (const listener of listeners) {
+        if (listener.length === 2) {
+          await new Promise<void>((resolve) => {
+            listener(data, resolve)
+          })
+        } else {
+          listener(data)
```
Comment (jochem-brouwer) [inline packages/evm/src/evm.ts:231]: This is the async logic handler now, right? This is not part of `eventemitter3`?

File: packages/evm/test/eips/eip-5656.spec.ts [hunk 7]
```diff
@@ -91,11 +91,13 @@ describe('should test mcopy', () => {
...
       let currentMem = ''
 
-      evm.events.on('step', (e) => {
+      evm.events.on('step', (e, resolve) => {
         if (e.opcode.name === 'STOP') {
           currentMem = bytesToHex(e.memory)
           assert.equal(currentMem, '0x' + situation.post, 'post-memory correct')
         }
+        // we need to call resolve() to avoid the event listener hanging
+        resolve?.()
```
Comment (jochem-brouwer) [inline packages/evm/test/eips/eip-5656.spec.ts:100]: But we can also remove the `resolve` here, right? (So no change in this file)?

File: packages/evm/test/runCall.spec.ts [hunk 8]
```diff
@@ -627,11 +627,12 @@ describe('RunCall tests', () => {
 
     let verifyMemoryExpanded = false
 
-    evm.events.on('step', (e) => {
+    evm.events.on('step', (e, resolve) => {
       assert.ok(e.memory.length <= 96)
       if (e.memory.length > 0) {
         verifyMemoryExpanded = true
       }
+      resolve?.()
```
Comment (jochem-brouwer) [inline packages/evm/test/runCall.spec.ts:635]: Same here, remove the resolve from the listener and the call?

File: packages/vm/README.md [hunk 9]
```diff
@@ -380,6 +379,22 @@ You can subscribe to the following events:
...
 
+Note, if subscribing to events with an async listener, specify the second parameter of your listener as a `resolve` function that must be called once your listener code has finished.
+
+```ts
+// ./examples/eventListener.ts#L10-L18
+
+// Setup an event listener on the `afterTx` event
+vm.events.on('afterTx', (event, resolve) => {
+  console.log('asynchronous listener to afterTx', bytesToHex(event.transaction.hash()))
+  resolve?.()
```
Comment (jochem-brouwer) [inline packages/vm/README.md:390]: An explicit note that `resolve` must be called would be nice here :)

**Review by jochem-brouwer — COMMENTED** [review]
Looks clean and good, however, I see that for VM and EVM there are now changes where it seems that even on sync events `resolve()` is added and called (in the tests). These should be removed.

Also, the "breaking change" here is that blockchain events do now not support async events. Not sure if we want to put that back in, but if anyone is listening async on `deletedCanonicalBlocks` we definitely should.

I will also test this ASAP :)

File: packages/evm/README.md (hunk 5 above)
Comment (acolytec3) [reply packages/evm/README.md:85]: yes

File: packages/client/test/net/protocol/rlpxsender.spec.ts (hunk 3 above)
Comment (acolytec3) [reply packages/client/test/net/protocol/rlpxsender.spec.ts:63]: Right, they aren't directly related but they're pointless tests and don't actually test anything other than that when you try to access a non-existent property on an object that Javascript throws.  I just stumbled across it because it referenced `EventEmitter`.

File: packages/evm/src/evm.ts (hunk 6 above)
Comment (acolytec3) [reply packages/evm/src/evm.ts:231]: Correct, if the listener has 2 arguments, we assume it's looking for the event object and then the resolve function (and therefore is async).  Otherwise, we just call the listener.

File: packages/client/src/sync/fullsync.ts (hunk 1 above)
Comment (acolytec3) [reply packages/client/src/sync/fullsync.ts:74]: Will make this change.  Not really sure why we even propagate this other than for debug logging.

File: packages/client/src/types.ts (hunk 2 above)
Comment (acolytec3) [reply packages/client/src/types.ts:55]: `eventEmitter3` doesn't need this additional layer of type expansions and we can just use the basic `EventParams` interface keyed by the properties of the `Event` enum.

File: packages/client/test/net/protocol/rlpxsender.spec.ts (hunk 4 above)
Comment (acolytec3) [reply packages/client/test/net/protocol/rlpxsender.spec.ts:46]: I've reversed this edit.  For some reason, Typescript wasn't paying nicely with it before.

**Review by scorbajio — APPROVED** [review]
LGTM

File: packages/client/src/sync/sync.ts [hunk 10]
```diff
@@ -145,9 +145,9 @@ export abstract class Synchronizer {
 
   abstract syncWithPeer(peer?: Peer): Promise<boolean>
 
-  resolveSync(height?: number) {
+  resolveSync(height?: bigint) {
     this.clearFetcher()
-    const heightStr = typeof height === 'number' && height !== 0 ? ` height=${height}` : ''
+    const heightStr = typeof height === 'bigint' && height !== BIGINT_0 ? ` height=${height}` : ''
```
Comment (holgerd77) [inline packages/client/src/sync/sync.ts:150]: Do these things even solve any bugs?? Or did you just also change the type coming in?
